executable-spec/src/Cardano/Ledger/Spec/STS/Update/Data.hs

```diff
@@ -1,29 +1,130 @@
 
 module Cardano.Ledger.Spec.STS.Update.Data where
 
-import Data.Word (Word64)
+--import Data.Word (Word64)
 import           Data.Monoid.Generic (GenericMonoid (GenericMonoid),
                      GenericSemigroup (GenericSemigroup))
 import           Data.Set (Set)
 import           GHC.Generics (Generic)
-
+import           Data.Text (Text)
+import           Numeric.Natural (Natural)
+import           Data.Hashable (Hashable)
+import qualified Data.Hashable as H
 import qualified Ledger.Core as Core
+import           Data.Map.Strict (Map)
+
+
+-- |A unique ID of a software update
+newtype UpId = UpId { getUpId :: Integer --Word64
+                    }
+  deriving (Show, Eq, Ord, Hashable, Generic)
+
+instance Core.HasHash (UpId) where
+  hash (UpId i) = Core.Hash $ H.hash i
+
+instance Num UpId where
+  (+) a b = UpId $ (getUpId a) + (getUpId b)
+
+-- | Protocol version
+data ProtVer = ProtVer
+  { _pvMaj :: Natural
+  , _pvMin :: Natural
+  , _pvAlt :: Natural
+  } deriving (Eq, Generic, Ord, Show, Hashable)
+
+-- | Application version
+newtype ApVer = ApVer Natural
+  deriving stock (Generic, Show)
+  deriving newtype (Eq, Ord, Num, Hashable)
+
+-- | Consensus Protocol Parameter Name
+data ParamName =
+    BlockSizeMax
+  | TxSizeMax
+  | SlotSize
+  | EpochSize
+  deriving (Eq, Generic, Ord, Show, Hashable)
+
+-- | Metadata structure for SIP
+data SIPMetadata =
+  SIPMetadata { versionFrom :: (ProtVer, ApVer)
+                -- ^ The version the this SIP has been based on
+              , versionTo :: (ProtVer, ApVer)
+              -- ^ the version after the SIP takes effect
+              , impactsConsensus :: Bool
+              -- ^ Flag to determine an impact on the underlying consensus protocol
+              , impactsParameters :: [ParamName]
+              -- ^ List of protocol parameters impacted
+              } deriving (Eq, Generic, Ord, Show, Hashable)
+
+-- | Contents of a SIP
+data SIPData =
+  SIPData {  url :: Text
+            -- ^ URL pointing at the SIP
+          , metadata :: SIPMetadata
+            -- ^ SIP Metadata
+          }
+  | NullSIPData
+  deriving (Eq, Generic, Ord, Show, Hashable)
+
+instance Core.HasHash (SIPData) where
+  hash a = Core.Hash $ H.hash a
 
 -- | System improvement proposal
 data SIP =
-  SIP { id :: Word64
+  SIP { id :: UpId
       -- ^ Submission proposal id.
+      , sipHash :: Core.Hash
+      -- ^ Hash of the SIP contents (`SIPData`)
       , author :: Core.VKey
       -- ^ Who submitted the proposal.
+      , salt :: Int
+      -- ^ The salt used during the commit phase
       }
+  deriving (Eq, Generic, Ord, Show, Hashable)
+
+instance Core.HasHash (SIP) where
+  hash a = Core.Hash $ H.hash a
+
+-- | A commitment data type.
+-- It is the `hash` $ salt ++ sip_owner_pk ++ `hash` `SIP`
+newtype Commit = Commit { getCommit :: Core.Hash}
+  deriving (Show, Eq, Ord)
+
+-- | The System improvement proposal at the commit phase
+data SIPCommit =
+  SIPCommit { _commit :: Commit
```

Comment:
We don't need to prepend underscores in the field names (I don't think we want to use lenses).

Reply:
OK. I will change it.

Reply:
Later on, in the state, we have a `Map Commit SIpCommit`. Do we need to have a `Commit` field here?

Reply:
It is the access key for the look up. We look up in the state for a commit value that equals the one calculated from the revealed SIP. This matching is the basic rule for a reveal to be successful.

Reply:
I have removed the underscore, except from the author field, which appears in two data types.